source/versioned-api/versioned-api.rst

```diff
@@ -0,0 +1,273 @@
+:Author: Andreas Braun
+:Advisors: Jeff Yemin, A. Jesse Jiryu Davis, Patrick Freed, Oleg Pudeyev
+:Status: Draft
+:Type: Standards
+:Minimum Server Version: N/A
+:Last Modified: 2020-11-05
+
+.. contents::
+
+--------
+
+Abstract
+========
+
+As MongoDB moves toward more frequent releases (a.k.a. continuous delivery), we
+want to enable users to take advantage of our rapidly released features, without
+exposing applications to incompatible server changes due to automatic server
+upgrades. A versioned API will help accomplish that goal.
+
+
+META
+====
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in `RFC 2119 <https://www.ietf.org/rfc/rfc2119.txt>`__.
+
+This document tends to use "SHOULD" more frequently than other specifications,
+but mainly in the context of providing guidance on writing test files. This is
+discussed in more detail in `Design Rationale`_.
+
+
+Specification
+=============
+
+Background
+----------
+
+When applications interact with MongoDB, both the driver and the server
+participate in executing operations. Therefore, when determining application
+compatibility with MongoDB, both the driver and the server behavior must be
+taken into account.
+
+An application can specify the server API version when creating MongoClient.
+When this is done:
+
+- The client sends the specified API version to the server, causing the server
+  to behave in a manner compatible with that API version.
+- The driver will behave in a manner compatible with the driver's behavior as of
+  the release when the driver first started to support the specified server API
+  version.
+
+Presently there is no specification for how a driver must behave when a
+particular server API version is requested, or what driver operations are
+subject to API compatibility guarantees. Such requirements may be stipulated in
+subsequent specifications.
+
+This specification requires MongoClient to validate that it supports the
+specified server API version, if any, but does not define what such support
+means.
+
+
+``MongoClient`` changes
+-----------------------
+
+``MongoClient`` instances accept a new ``serverApi`` option to allow the user to
+declare an API version:
+
+.. code:: typescript
+
+   class MongoClient {
+       MongoClient(... serverApi: ServerApi);
+   }
+
+   enum ServerApiVersion {
+       v1 = "1",
+   }
+
+   class ServerApi {
+       version: string|ServerApiVersion;
+       strict: Optional<Boolean>; // Default false
+       deprecationErrors: Optional<Boolean>; // Default false
+   }
+
+Drivers SHOULD group the ``serverApi`` option with other similar client options
+like ``autoEncryptionOpts``.
+
+
+ServerApiVersion enumeration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This specification and subsequent specifications will define the known API
+versions. Drivers SHOULD define an enumeration containing the known API
```

Comment:
One issue with using an enum for this is that a lot of languages allow exhaustive `switch` statements over the various cases, meaning adding a new case constitutes a breaking change (Rust doesn't have this problem, by the way). If we intend to add support for new API versions in non-major releases, then we might want to include the addition requirement here that "drivers MUST be able to introduce new versions without breaking source compatibility" or something like that.

Reply:
Technically, drivers don't have to add compatibility for a new API Version in the current major release series, but can do so in a new major release if that's required. I've asked the leads for input on how their languages treat this to make an informed decision and will revisit this once I have some answers.

Reply:
After discussion we've come to the conclusion that it makes sense to let drivers handle this individually without prescribing a one-size-fits-all solution. For example, a new enum value could produce a warning in C/C++ (via `-Wswitch-enum` in gcc), which in turn can cause compiler errors if warnings are treated as errors. However, since it is the user's choice to treat this strictly it is up to them to take care of this and such a change would be allowed in a minor release. Swift is a different story, as users may want to work with the enumeration in a `switch` statement, so adding a new version could be a BC break.

With this in mind, I'd suggest adding the following wording:
> Drivers MUST ensure that adding new API versions to this enumeration does not result in backward compatibility breaks in non-major releases. This can be the case in languages that allow exhaustive ``switch`` statements (e.g. Swift). In these cases, drivers should either refrain from using enumerations or only introduce new API versions in major releases.

This ensures that readers of the spec are aware of the potential issue, but can decide how they want to deal with it: either by not using enumerations (as is done in the Swift driver in some instances), by choosing to not treat it as a BC breaking change (as the C/C++ driver might want to do), or by avoiding the issue altogether and releasing a new major version.

Reply:
I'm fine with the decision to provide flexibility, but I don't think we should mention the option of a major release just for a new API version, which likely otherwise wouldn't require any other driver changes except the addition of the enumeration.  Major releases are painful for users and require a lot of planning, and that's precisely what we are trying to avoid.

Reply:
I would just remove the last sentence.

Reply:
Yeah, I agree. Updating the default version API should certainly require a major release (imo), but simply adding support  for a new one shouldn't require one. The rest of the wording lgtm.

Reply:
Removed the last sentence as per your suggestions.

Reply:
lgtm, but leaving for patrick to resolve